### ISSUE TITLE
Expose table alterations under `alter` namespace

### ISSUE THREAD

**Issue by ion-elgreco** [delta-io/delta-rs#1909]
# Description

**Use Case**
Eventually, we will have multiple alterations possible on the table, such as setting/unsetting table properties, adding and removing columns and so forth. We can cluster these nicely together under a single namespace called `alter`. The API will look like this:

```python
DeltaTable().alter.set_table_properties()
DeltaTable().alter.unset_table_properties()
DeltaTable().alter.add_columns()
DeltaTable().alter.change_columns()
DeltaTable().alter.replace_columns()
DeltaTable().alter.add_constraints()
DeltaTable().alter.drop_constraints()

```
**Related Issue(s)**
https://github.com/delta-io/delta-rs/issues/1663

**Comment by roeap** [comment]
In principle I have no too strong feeling about bundling some commands under a common property, much like we do for optimize.

Given the name `alter` though, I would suggest restricting it to things things that can be done via the `ALTER TABLE` command, as we may want / need to implement that operation at some point.

The way things seem to be going with the Delta Protocol, it seems table features are front and center when it comes to configuring tables. Along with that some configuration is becoming more complex. As such we may consider exposing `set_table_properties` as a low level (discouraged) API only and instead model this around table features. Something along the lines of

```py
def enable_feature(name: FeatureName, config: dict)...
```

The advantage may be that is is easier for us to validate the configuration as configuration for specific features may include multiple keys that need to be consistent and (as far as I understand) may even require setting domain metadata at some point.

**Comment by ion-elgreco** [comment]
@roeap it's mostly inspired from here the SQL alter operations: https://docs.databricks.com/en/sql/language-manual/sql-ref-syntax-ddl-alter-table.html

I am not entirely following you on why set_table_proprties should be a low-level API. Because not every table property belongs to a certain feature, right? : https://books.japila.pl/delta-lake-internals/DeltaConfigs/#appendOnly

**Comment by roeap** [comment]
yes, there is config that is unrelated to features... mainly saying that the config that is related to features should maybe modeled as such ...

**Comment by ion-elgreco** [comment]
@roeap I am going to start looking to this soon, just to be clear. For configs that are related to features, should we raise when someone tries to add or remove them in set table propeties way?

**Comment by dtheodor** [comment]
Does this issue cover adding support for DDL statements in general, such as [CREATE TABLE](https://docs.databricks.com/en/sql/language-manual/sql-ref-syntax-ddl-create-table-using.html) and [ALTER TABLE ...](https://docs.databricks.com/en/sql/language-manual/sql-ref-syntax-ddl-alter-table.html)? Currently only possible with spark.

**Comment by ion-elgreco** [comment]
> Does this issue cover adding support for DDL statements in general, such as [CREATE TABLE](https://docs.databricks.com/en/sql/language-manual/sql-ref-syntax-ddl-create-table-using.html) and [ALTER TABLE ...](https://docs.databricks.com/en/sql/language-manual/sql-ref-syntax-ddl-alter-table.html)? Currently only possible with spark.

Create table is already covered with the `create` operation.

Some alter operations are already available, this is still a work in progress to add more, such as `add columns` operation

**Comment by dtheodor** [comment]
Complete support for alter operations would make this project useful for lightweight migrations, omitting the need for a spark cluster to perform them

**Comment by zero-element** [comment]
Currently, there is no `drop_columns` here. Does anyone know the background or the challenges behind implementing it?

**Comment by ion-elgreco** [comment]
@zero-element I have a draft PR but it's a bit more complicated since it also requires column mapping. So if you want to add support for this, we first need column mapping support

**Comment by zero-element** [comment]
> [@zero-element](https://github.com/zero-element) I have a draft PR but it's a bit more complicated since it also requires column mapping. So if you want to add support for this, we first need column mapping support

I have noticed the issue at https://github.com/delta-io/delta-rs/issues/930. Do you know if anyone is working or planning to work on this? Are there any suggestions on how to start with this?

**Comment by ion-elgreco** [comment]
@zero-element there is no one working on column mapping atm. You would need to understand how column mapping works under the hood and interacts with the internal apis. It requires integration with delta-kernel-rs for this and you would need to make the reading/scanning and writing compatible with this.

If you really want this feature, I would suggest looking also at how it's implemented in delta-spark so you get some idea's and understand how you could achieve the same result with our internal rust apis :)

**Comment by zero-element** [comment]
@ion-elgreco Thank you for your suggestions. In the near term, I may not have enough time to delve deeper into such a complex issue. However, if there is no progress within a few months, I will give it a try. BTW, it seems that many critical functionalities rely on column mapping (such as various alter operations and deletion vectors), will this be an important direction moving forward?

**Comment by ion-elgreco** [comment]
@zero-element that's correct, column mapping is needed for some of these operations. The only way for this to become supported is that someone with enough interest works on it, or someone gets sponsored to do it. Neither of which has happened atm : P

**Comment by ArulJerald** [comment]
take
